File: tools/xstore/core/engine/store.py
# Copyright 2021 Alibaba Group Holding Limited.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import ClassVar

from core import Context
from .engine import Engine

_engine_cls_map = {}


def register_engine_class(engine: str, cls: ClassVar[Engine]):
    if engine in _engine_cls_map:
        raise RuntimeError('duplicate engine_name: ' + engine)
    _engine_cls_map[engine] = cls


def new_engine(engine: str, context: Context) -> Engine:
    """
    Construct a new name object of target engine_name.

    :param engine: engine_name name.
    :param context: current context.
    :return:
    """
    engine_cls = _engine_cls_map[engine]
    if not engine_cls:
        raise ValueError('engine_name not supported: ' + engine)
    return engine_cls(context)

File: pkg/util/config/store/store.go
/*
Copyright 2021 Alibaba Group Holding Limited.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package store

import "sync/atomic"

type Store struct {
	val atomic.Value
}

func (s *Store) Get() interface{} {
	return s.val.Load()
}

func (s *Store) Update(val interface{}) interface{} {
	return s.val.Swap(val)
}

func (s *Store) Cached(lazy bool) *CachedStore {
	if lazy {
		return &CachedStore{
			Store:  s,
			valPtr: nil,
		}
	} else {
		val := s.Get()
		return &CachedStore{
			Store:  nil,
			valPtr: &val,
		}
	}
}

type CachedStore struct {
	*Store
	valPtr *interface{}
}

func (s *CachedStore) Get() interface{} {
	if s.valPtr != nil {
		return *s.valPtr
	}
	val := s.Store.Get()
	s.valPtr = &val
	return *s.valPtr
}

func (s *CachedStore) Cached(lazy bool) *CachedStore {
	if lazy || s.valPtr != nil {
		return s
	} else {
		val := s.Store.Get()
		return &CachedStore{
			Store:  nil,
			valPtr: &val,
		}
	}
}
